nexus/db-queries/src/db/datastore/pub_test_utils.rs

```diff
@@ -12,32 +12,132 @@ use crate::authz;
 use crate::context::OpContext;
 use crate::db;
 use crate::db::DataStore;
-use dropshot::test_util::LogContext;
 use omicron_test_utils::dev::db::CockroachInstance;
+use slog::Logger;
 use std::sync::Arc;
 use uuid::Uuid;
 
+#[cfg(test)]
+mod test {
+    use super::*;
+    use nexus_test_utils::db::test_setup_database;
+
+    enum TestState {
+        Pool { pool: Arc<db::Pool> },
+        RawDatastore { datastore: Arc<DataStore> },
+        Datastore { opctx: OpContext, datastore: Arc<DataStore> },
+    }
+
+    /// A test database with a pool connected to it.
+    pub struct TestDatabase {
```

Comment:
Might be missing something but can this go in nexus-test-utils?

Reply:
If it's cool with you I'm going to give this a shot in the follow-up PR: https://github.com/oxidecomputer/omicron/pull/6893

Reply:
Totally. thanks.

Reply:
Okay, I have an answer here: pulling this test utility out of `nexus-db-queries` is a nightmare from a circular dependency point-of-view.

If I move the `TestDatabase` into `nexus-test-utils`, it contains a `Datastore` object from `nexus-db-queries`.

Under test, `nexus-db-queries` wants to use a `TestDatabase`, which means we have a dependency ordering of:

`nexus-db-queries` -> `nexus-test-utils` (to get the TestDatabase) -> `nexus-db-queries` (to get the datastore)

Then, if any tests in `nexus-db-queries` try to use the `DataStore`, within `TestDatabase`, it sees two incompatible versions of that struct.

Example:

```
 1  error[E0308]: mismatched types                                                                                                                                                                                                                                                                                                                                                                                                                                                          ▐
    --> nexus/db-queries/src/db/queries/external_ip.rs:948:46
     |
 948 |             LookupPath::new(self.db.opctx(), &**self.db.datastore())
     |             ---------------                  ^^^^^^^^^^^^^^^^^^^^^^ expected `datastore::DataStore`, found `DataStore`
     |             |
     |             arguments to this function are incorrect
     |
     = note: `DataStore` and `datastore::DataStore` have similar names, but are actually distinct types
 note: `DataStore` is defined in crate `nexus_db_queries`
    --> /home/smklein/repos/oxide/omicron/nexus/db-queries/src/db/datastore/mod.rs:182:1
     |
 182 | pub struct DataStore {
     | ^^^^^^^^^^^^^^^^^^^^
 note: `datastore::DataStore` is defined in the current crate
    --> nexus/db-queries/src/db/datastore/mod.rs:182:1
     |
 182 | pub struct DataStore {
     | ^^^^^^^^^^^^^^^^^^^^
     = note: the crate `nexus_db_queries` is compiled multiple times, possibly with different configurations
```

Reply:
Oh yeah that's a nightmare. Thanks for digging into this, and sorry for sending you on this wild goose chase!